### PR TITLE
Update `expr` matcher for Edition 2024 and add `expr_2021` nonterminal

### PR DESCRIPTION
**Pull request by eholk**
This commit adds a new nonterminal `expr_2021` in macro patterns, and `expr_fragment_specifier_2024` feature flag.

This change also updates `expr` so that on Edition 2024 it will also match `const { ... }` blocks, while `expr_2021` preserves the current behavior of `expr`, matching expressions without `const` blocks.

Joint work with @vincenzopalazzo.

Issue #123742

### CODE REVIEW

File: compiler/rustc_ast/src/token.rs [hunk 1]
```diff
@@ -905,6 +907,7 @@ impl NonterminalKind {
             },
             sym::pat_param => NonterminalKind::PatParam { inferred: false },
             sym::expr => NonterminalKind::Expr,
+            sym::expr_2021 if edition() >= Edition::Edition2024 => NonterminalKind::Expr2021,
```
Comment (eholk) [inline compiler/rustc_ast/src/token.rs:910]: From [the RFC](https://github.com/rust-lang/rfcs/blob/master/text/3531-macro-fragment-policy.md#:~:text=In%20the%20current%20edition%2C%20the%20next%20edition%2C%20and%20as%20many%20other%20editions%20as%20practical%2C%20add%20a%20new%20fragment%20specifier%20that%20preserves%20the%20behavior%20of%20the%20existing%20fragment%20specifier.):

> In the current edition, the next edition, and as many other editions as practical, add a new fragment specifier that preserves the behavior of the existing fragment specifier.

My interpretation of this is that we need to allow `expr_2021` on Edition 2021, and if practical also support it on Edition 2018 and 2015. I don't see any reason why we shouldn't be able to support this on all editions.

Comment (eholk) [reply compiler/rustc_ast/src/token.rs:910]: Actually, we probably can't do 2015 because `async` wasn't a keyword then.

Comment (vincenzopalazzo) [reply compiler/rustc_ast/src/token.rs:910]: **TL;DR:** It's safe to assume that `expr_2021` will work for every previous edition until we encounter another `expr_xx`, such as `expr_2015` (if this were the case).

One practical idea to support this range of editions is that `expr_2021` remains backwards compatible until `expr_xx`. If there is another `expr_xx`, then the compatibility range includes everything.

However, consider if someone uses the `async` keyword in an edition like 2015; the code will likely complain that `async` is not supported, right? Perhaps not, if the keyword is used only inside the macros.

Comment (eholk) [reply compiler/rustc_ast/src/token.rs]: This is fixed, as of [e0b7d92](https://github.com/rust-lang/rust/pull/123865/commits/e0b7d923b4900b382a2f20abb75a5aa000284542) we allow `expr_2021` on Edition 2021.

File: tests/ui/macros/expr_2021_inline_const.rs
Comment (fmease) [inline tests/ui/macros/expr_2021_inline_const.rs:1]: I wonder if there are any crazy edge cases to consider like nested MBEs where the innermost fragment specifier is a macro metavariable and decl macros come from different crates. E.g:

Upstream:

```rs
//@ edition: ???
#![feature(macro_metavar_expr)]

#[macro_export]
macro_rules! outer {
    ($frag:ident) => {
        macro_rules! inner {
            ($$it:$frag) => {}
        }
    }
}
```

Downstream:

```rs
//@ edition: ???

upstream::outer!(expr);
inner!(const { 0 });
```

I don't know if this yields anything interesting. I'm a bit tired right now, it's late. Just wanted to mention it.

**Comment by fmease** [comment]
I think the biggest question left is what exactly should be included in Rust 2024's `expr` as you've started to discuss in the tracking issue.

File: tests/ui/macros/expr_2021_inline_const.rs
Comment (eholk) [reply tests/ui/macros/expr_2021_inline_const.rs:1]: I've been tying my brain in knots about these kinds of scenarios recently, but I don't have good answers for them.

I think my approach would be to go ahead and merge this and then get wg-macros to start creating test cases to make sure what the behavior we have is and whether that's the behavior we want.

File: compiler/rustc_expand/src/mbe/quoted.rs [hunk 2]
```diff
@@ -92,6 +92,12 @@ pub(super) fn parse(
                                                 token::NonterminalKind::Ident
                                             },
                                         );
+                                    if kind == token::NonterminalKind::Expr2021
+                                        && !features.expr_fragment_specifier_2024
+                                    {
+                                        sess.dcx()
+                                            .emit_err(errors::Expr2021IsExperimental { span });
```
[the remark this answers is not in the document]
Comment (eholk) [reply compiler/rustc_expand/src/mbe/quoted.rs]: Thanks for the suggestion. This should be done now.

File: tests/ui/macros/expr_2021_old_edition.stderr [hunk 3]
```diff
@@ -0,0 +1,25 @@
+error: invalid fragment specifier `expr_2021`
```
[the remark this answers is not in the document]
Comment (eholk) [reply tests/ui/macros/expr_2021_old_edition.stderr:1]: I updated the help text so it should be better now. The way I did it was pretty clumsy though, so I'm open to suggestions for how to improve it.

**Comment by eholk** [comment]
@fmease - I addressed your simpler comments, so I think this should be ready for another look.

**Comment by vincenzopalazzo** [comment]
Looks like we should update the tests

```
  failures:
  
  ---- [ui] tests/ui/macros/expr_2021_inline_const.rs#edi2021 stdout ----
  diff of stderr:
  
  28	LL |     ($e:expr) => {
  29	   |      ^^^^^^^
  30	
  -	error: aborting due to 2 previous errors
  +	warning: the feature `inline_const` has been stable since 1.80.0-nightly and no longer requires an attribute to enable
  +	  --> $DIR/expr_2021_inline_const.rs:7:12
  +	   |
  +	LL | #![feature(inline_const)]
  +	   |            ^^^^^^^^^^^^
  +	   |
  +	   = note: `#[warn(stable_features)]` on by default
  +	
  +	error: aborting due to 2 previous errors; 1 warning emitted
  32	
  33	
```

File: tests/ui/macros/expr_2021_old_edition.stderr [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  --> $DIR/expr_2021_old_edition.rs:6:6
+   |
+LL |     ($e:expr_2021) => {
+   |      ^^^^^^^^^^^^
+   |
+   = help: fragment specifier `expr_2021` requires Rust 2021 or later
+           valid fragment specifiers are `ident`, `block`, `stmt`, `expr`, `pat`, `ty`, `lifetime`, `literal`, `path`, `meta`, `tt`, `item` and `vis`
```
Comment (fmease) [inline tests/ui/macros/expr_2021_old_edition.stderr:8]: I would probably split this into two separate `help` messages but er, doesn't matter

File: tests/ui/macros/expr_2021_inline_const.edi2021.stderr
Comment (fmease) [inline tests/ui/macros/expr_2021_inline_const.edi2021.stderr:1]: This (intentional) behavior is a bit unfortunate but it makes sense. Well, it makes the most sense when edition-dependent keywords are involved.

File: compiler/rustc_parse/src/parser/nonterminal.rs [hunk 5]
```diff
@@ -37,13 +38,19 @@ impl<'a> Parser<'a> {
...
                 // This exception is here for backwards compatibility.
                 && !token.is_keyword(kw::Let)
                 // This exception is here for backwards compatibility.
                 && !token.is_keyword(kw::Const)
             }
+            NonterminalKind::Expr => {
+                token.can_begin_expr()
+                // This exception is here for backwards compatibility.
+                && !token.is_keyword(kw::Let)
+                && (token.span.edition() >= Edition::Edition2024 || !token.is_keyword(kw::Const))
```
[the remark this answers is not in the document]
Comment (fmease) [reply compiler/rustc_parse/src/parser/nonterminal.rs:52]: not applied? :P

**Review by fmease — APPROVED** [review]
Thanks! My suggestion https://github.com/rust-lang/rust/pull/123865/files#r1577167279 wasn't applied, I assume that's an oversight because you've applied suggestions of mine that are similar to this.

Should be good to go after that, work like double-checking the behavior of weird nested macro decls can be done in follow-ups as suggested by you.

r=me then.

File: compiler/rustc_parse/src/parser/nonterminal.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (eholk) [reply compiler/rustc_parse/src/parser/nonterminal.rs:52]: Oh weird, I thought I had done this one.

File: tests/ui/macros/expr_2021_inline_const.rs
Comment (eholk) [reply tests/ui/macros/expr_2021_inline_const.rs:1]: I added a bullet about this to #123742 so we don't lose the suggestion.

File: tests/ui/macros/expr_2021_old_edition.stderr (hunk 4 above)
Comment (eholk) [reply tests/ui/macros/expr_2021_old_edition.stderr:8]: Yeah, I think that'd be better. Let me poke around at it some more, I'm still getting the hang of the diagnostics system.

**Comment by rylev** [comment]
@eholk @vincenzopalazzo @fmease the regression seems legit. It seems the primary benchmark affected by this `html5ever` seems to spend lots of time in `expand_crate` which could be impacted by this change. Thoughts on whether this needs further investigation or if we should just treat that particular benchmark as a quasi-stress-test and move on?

**Comment by vincenzopalazzo** [comment]
@rylev I was planning to discuss with @eholk and take a look at the code, idk if @eholk already gave a look at the code, but I need a couple of more days

**Comment by fmease** [comment]
Update: In #125389 I've checked the perf of a full revert. Doing a naive inversion / deduction, this PR "might've only had" 8 counts of secondary regressions (mean: +0.3%, range: [+0.5%, +0.3%]). Anyways, if I find the time I will look deeper into this. Otherwise, I'll just wait for https://github.com/rust-lang/rust/pull/123865#issuecomment-2122672475.

File: compiler/rustc_parse/src/parser/nonterminal.rs [hunk 6]
```diff
@@ -37,13 +37,19 @@ impl<'a> Parser<'a> {
...
                 // This exception is here for backwards compatibility.
                 && !token.is_keyword(kw::Let)
                 // This exception is here for backwards compatibility.
                 && !token.is_keyword(kw::Const)
             }
+            NonterminalKind::Expr => {
+                token.can_begin_expr()
+                // This exception is here for backwards compatibility.
+                && !token.is_keyword(kw::Let)
+                && (token.span.edition().at_least_rust_2024() || !token.is_keyword(kw::Const))
```
Comment (vincenzopalazzo) [inline compiler/rustc_parse/src/parser/nonterminal.rs:51]: I am not a performance expert, so I might be wrong here, but I think the only thing that could cause a performance issue is the token.span.edition().

If we follow the call stack, the edition() will resolve in the following call: https://github.com/rust-lang/rust/blob/master/compiler/rustc_span/src/hygiene.rs#L877-L879.

If we repeat this call for every expression token, we could see a performance decrease.

**Review by vincenzopalazzo — COMMENTED** [review]
I finally had a look at the code and I post my analysis, I am not sure how to solve this BTW :/

File: tests/ui/macros/expr_2021_inline_const.rs
Comment (eholk) [reply tests/ui/macros/expr_2021_inline_const.rs:1]: We have a PR with that adds a test case around this scenario: https://github.com/rust-lang/rust/pull/129755

Not entirely sure what the correct behavior should be, but it seems like the behavior we have now isn't bad.

**Pull request merged**